compiler/rustc_passes/src/dead.rs

```diff
@@ -802,7 +802,7 @@ fn check_item<'tcx>(
             // And we access the Map here to get HirId from LocalDefId
             for local_def_id in local_def_ids {
                 // check the function may construct Self
-                let mut may_construct_self = true;
+                let mut may_construct_self = false;
                 if let Some(fn_sig) =
                     tcx.hir().fn_sig_by_hir_id(tcx.local_def_id_to_hir_id(local_def_id))
                 {
```

Comment:
Is there a test for the case where a method could construct `Self`?

Reply:
There was a comment linking to https://github.com/rust-lang/rust/blob/master/tests/ui/lint/dead-code/unused-pub-struct.rs but got deleted. Reposting it here.